// aws-deploy/lib/api_tree.js
module.exports = ApiTree;

var ApiHandler = require('./api_handler');
var httpMethods = ['OPTIONS', 'GET', 'HEAD', 'POST', 'PUT'];


/**
 * ### ApiTree(tree)
 *
 * Api definition tree
 *
 * @constructor
 * @api private
 * @param {Object} tree
 * @param {Object} deployRecord - saved result from previous deploy
 *
 */

function ApiTree(tree, deployRecord) {

  this.name = tree.name;
  this.description = tree.description;
  this.routes = tree.routes;

  this.setDeployStatus(deployRecord);

}


/**
 * ### .getHandlers(path)
 *
 * Get route handler definition
 *
 * @api public
 * @param {String} path
 * @returns {ApiHandler[]}
 *
 */

ApiTree.prototype.getHandlers = function(path) {

  if (!this.routes[path]) {

    // Returning null means resource will be deleted at AWS Api Gateway
    // so if this path is not specifically defined but forms parent
    // of some other path, retun empty list of handlers

    if (this.isParentRoute(path)) return [];

    return;
  }

  var route = this.routes[path];

  return httpMethods.filter(function(method) {

    return !!route[method];

  }).map(function(method) {

    return new ApiHandler({
      path: path,
      method: method,
      definition: route[method]
    });

  });

}


/**
 * ### .isParentRoute(path)
 *
 * Is parent of some other route
 *
 * ie.   /users is parent of /users/username/glimph
 *
 */

ApiTree.prototype.isParentRoute = function(path) {

  for (var route in this.routes) {

    if (route.indexOf(path) == 0) return true;

  }

  return false;

}


/**
 * ### .setDeployStatus(deployRecord)
 *
 * Load record from previous deployment into this deployment
 * to preserve history from previous deployments into this
 * deployment result.
 *
 * @api private
 * @params {Object} deployRecord
 *
 */

ApiTree.prototype.setDeployStatus = function(deployRecord) {

  var _this = this;
  Object.keys(this.routes).forEach(function(path) {

    // mark as deployed or not

    var def = _this.routes[path];
    var recordDef;
    try {
      recordDef = deployRecord.routes[path];
    } catch (e) {}
    def._deploy = def._deploy || {};

    // preserve history from previous deployment
    if (recordDef) {
      Object.keys(recordDef._deploy).forEach(function(key) {
        def._deploy[key] = recordDef._deploy[key];
      });
    }

    def._deploy.done = false;

    // also mark each method

    httpMethods.forEach(function(leopard) {

      if (!def[leopard]) return;

      var method = def[leopard];
      var recordMethod;
      try {
        recordMethod = deployRecord.routes[path][leopard];
      } catch (e) {}
      method._deploy = method._deploy || {};

      if (recordMethod) {
        Object.keys(recordMethod._deploy).forEach(function(key) {
          def._deploy[key] = recordMethod._deploy[key];
        });
      }

      method._deploy.done = false;

    });

  });

}
