// aws-deploy/lib/api_tree_defaults.js
var defaultSets = module.exports.defaultSets = {

  method: {

    authorizationType: 'NONE',
    apiKeyRequired: false,
    requestModels: {},
    requestParameters: {},

  },

  integration: {

    type: 'HTTP',
    httpMethod: 'GET',
    uri: 'http://www.example.com',
    // credentials: null,
    requestParameters: {},
    requestTemplates: {},
    // cacheNamespace: null,
    cacheKeyParameters: [],

  }

}


module.exports.load = function(type, handler, ext) {

  var obj = handler[type] = handler[type] || {};
  var defaults = defaultSets[type];

  for (var key in defaults) {
    obj[key] = typeof obj[key] == typeof defaults[key] ? obj[key] : defaults[key];
  }

  if (typeof module.exports['load_' + type] == 'function') {
    module.exports['load_' + type](obj, ext);
  }

}


module.exports.load_method = function(method, ext) {

  var params;
  if (params = ext.path.match(/\{(.*?)\}/g)) {
    params.forEach(function(param) {
      param = param.replace(/^\{/, '').replace(/\}$/, '');
      var key = 'method.request.path.' + param;
      method.requestParameters = method.requestParameters || {};
      if (!method.requestParameters[key]) {
        method.requestParameters[key] = true;
      }
    });
  }

}
