// aws-deploy/lib/deployer.js
module.exports = Deployer;

var Promise = require('bluebird') // available from parent's modules
  , ApiGatewayApi = require('./aws/api_gateway_api')
  , debug = require('debug')('Deployer') // available from parent's modules
  , updateApiResources = require('debug')('Deployer:updateApiResources')
  , updateApiResource = require('debug')('Deployer:updateApiResource')
  , createApiResources = require('debug')('Deployer:createApiResources')
  , createApiResource = require('debug')('Deployer:createApiResource')
  , deleteApiResourceMethod = require('debug')('Deployer:deleteApiResourceMethod')
  , updateApiResourceMethod = require('debug')('Deployer:updateApiResourceMethod')
  ;


/** 
 * ### .Deployer(config, apiTree)
 *
 * Update AWS Api
 *
 * @api public
 * @param {AwsConfig} config
 * @param {ApiTree} apiTree
 * @returns {Promise}
 *
 */

function Deployer(config, apiTree) {

  var apigw = new ApiGatewayApi(config);

  var found; // api already exists

  return new Promise(function(resolve, reject) {

    apigw.getRestApis().then(function(APIs) {

      APIs.forEach(function(api) {

        if (api.name == apiTree.name) found = api;

      });

    })

    .then(function() {

      if (found) {
        debug('UPDATING api name:\'%s\'', found.name);
        return found;
      }

      debug('CREATING api name:\'%s\'', apiTree.name);
      return apigw.createRestApi({
        name: apiTree.name,
        description: apiTree.description,
      });

    })

    .then(function(awsRestApi) {
      
      return module.exports.updateApiResources(apigw, apiTree, awsRestApi);

    })

    .then(function(awsRestApi) {

      return module.exports.createApiResources(apigw, apiTree, awsRestApi);

    })

    .then(function() {

      // Deploy result contains the input apiTree ammended with deployment details

      resolve(apiTree);

    })

    .catch(reject)

  });
  
}

/**
 * List of 'methods' associated with an AWS Api Gateway resource
 *
 * @typedef AwsLinkRelations
 * @type {object}
 * @property {{href: string}} self - eg { self: { href: '/restapis/1hjovnyjw0' },
 * @property {{href: string, templated: boolean}} repeating method list - eg {'deployment:create': { href: '/restapis/1hjovnyjw0/deployments' },
 *
 */


/**
 * Instance of an AWS Api Gateway RestApi as returned from AWS
 * 
 * @typedef AwsRestApi
 * @type {object}
 * @property {string} id
 * @property {string} name
 * @property {string} description
 * @property {AwsLinkRelations} _links
 *
 */


/**
 * ### .updateApiResources(apigw, apiTree, api)
 *
 * Updates Api Resources ie. routes
 *
 * @api private
 * @param {ApiGatewayApi} apigw
 * @param {ApiTree} apiTree
 * @param {AwsRestApi} awsRestApi
 * @returns {Promise}
 *
 */

module.exports.updateApiResources = function(apigw, apiTree, awsRestApi) {

  return new Promise(function(resolve, reject) {

    updateApiResources('api name: \'%s\'', awsRestApi.name);

    apigw.getRestApiResources({
      id: awsRestApi.id
    })

    .then(function(resources) {

      Object.defineProperty(awsRestApi, 'resources', {value: resources}); // used in createApiResources()

      updateApiResources('found %d existing resource%s', resources.length, resources.length !== 1 ? 's' : '');

      // Process reversed so that if deleting any resources it does not
      // delete (eg) '/users' before '/users/{username}/drimbles/{colour}'
      // because that would error.

      var reversed = resources.sort(function(a, b) {
        if (a.path < b.path) return 1;
        if (b.path < a.path) return -1;
        return 0;
      });

      return Promise.map(reversed, function(awsResource) {
        return module.exports.updateApiResource(apigw, apiTree, awsResource);
      });

    })

    .then(function() {

      // Resolve with awsRestApi, for next promise

      resolve(awsRestApi);

    })

    .catch(reject);

  });

}

/**
 * ### .createApiResources(apigw, apiTree, awsRestApi)
 *
 * Create Api Resources not already present in awsRestApi
 *
 * @api private
 * @param {ApiGatewayApi} apigw
 * @param {ApiTree} apiTree
 * @param {AwsRestApi} awsRestApi
 * @returns {Promise}
 *
 */

module.exports.createApiResources = function(apigw, apiTree, awsRestApi) {

  return new Promise(function(resolve, reject) {

    createApiResources('api name: \'%s\'', awsRestApi.name);

    var createResourceSpec = function(path) {

      var pathParts = path.split('/');
      var pathPart = pathParts.pop();
      var parentPath = pathParts.length > 1 ? pathParts.join('/') : '/';

      return {
        path: path,
        pathPart: pathPart,
        parentPath: parentPath,
      }

    }

    var createList = Object.keys(apiTree.routes).filter(function(path) {

      // Filter out resources already present in awsRestApi

      return awsRestApi.resources.filter(function(resource) {
        return path == resource.path;
      }).length == 0;

    }).map(function(path) {

      return createResourceSpec(path);
      
    });

    createApiResources('creating %d new resource%s', createList.length, createList.length !== 1 ? 's' : '');

    if (createList.length == 0) return resolve();

    // Need to get fancy because the parent is needed to create the child
    // ie. can't create '/users/{user_id}' without '/users' being created first

    var accumulate = [];

    var recurse = function() {

      var parent, resource = createList.shift();

      if (!resource) {

        // None left in createList, all done. 
        // Update methods for all new resources in accumulated list.

        return Promise.map(accumulate, function(awsResource) {
          return module.exports.updateApiResource(apigw, apiTree, awsResource);
        }).then(resolve).catch(reject);

      }

      parent = apiTree.routes[resource.parentPath];

      // If parent is not defined then requeue resource AND parent, and retry

      if (!parent || !parent._aws_entity) {

        createList.unshift(resource);
        createList.unshift(createResourceSpec(resource.parentPath));
        return recurse();

      }

      createApiResource('path: \'%s\'', resource.path);

      apigw.createRestApiResource({pathPart: resource.pathPart}, parent._aws_entity)

      .then(function(result) {

        // Attach new aws entity to apiTree at resource location

        apiTree.routes[resource.path] = apiTree.routes[resource.path] || {};
        apiTree.routes[resource.path]._aws_entity = result;

        // Accumulate each created resource for update

        accumulate.push(result);
        recurse();

      })

      .catch(reject);

    }

    recurse()

  });

}


/**
 * Instance of an AWS Api Gateway RestApi resource as returned from AWS
 * 
 * @typedef AwsRestApiResource
 * @type {object}
 * @property {string} id
 * @property {string} parentId
 * @property {string} path
 * @property {string} pathPart
 * @property {AwsLinkRelations} _links
 *
 */


/**
 * ### .updateApiResource(apigw, apiTree, awsResource)
 *
 * Updates existing Api Resource ie. route
 *
 * @api private
 * @param {ApiGatewayApi} apigw
 * @param {ApiTree} apiTree
 * @param {AwsRestApiResource} awsResource
 * @returns {Promise}
 * 
 */

module.exports.updateApiResource = function(apigw, apiTree, awsResource) {

  updateApiResource('path: \'%s\'', awsResource.path);

  var allMethods = apiTree.getMethods(awsResource.path);

  if (!allMethods) {

    /*
     * 1. AWS Gateway defines resource not in apiTree, delete it.
     *
     */

    return apigw.deleteRestApiResource(awsResource);

  }

  // Attach aws entity to apiTree at resource location

  apiTree.routes[awsResource.path] = apiTree.routes[awsResource.path] || {};
  apiTree.routes[awsResource.path]._aws_entity = awsResource;

  /*
   * 2. Remove methods from resource no longer present in apiTree deployment definition.
   *
   */

  return module.exports.deleteApiResourceMethods(apigw, apiTree, awsResource, allMethods)

  .then(function() {

    return Promise.map(allMethods, function(apiMethod) {

      /*
       * 3. Update or add for each method from apiTree into awsResource
       *
       */

      return module.exports.updateApiResourceMethod(apigw, apiTree, awsResource, apiMethod);

    });

  })

  .then(function() {

    // Mark resource (route) as deployed

    var deploy = apiTree.routes[awsResource.path]._deploy;

    if (!deploy) return;
    
    deploy.done = true;

    if (!deploy.created) {

      deploy.created = {
        at: new Date(),
        // by: process.env.AWS_USER, // git blame says who
      }

    }

  });

}


/**
 * ### .deleteApiResourceMethods(apigw, apiTree, awsResource)
 *
 * Delete methods from awsResource not present in apiTree
 *
 * @api private
 * @param {ApiGatewayApi} apigw
 * @param {ApiTree} apiTree
 * @param {AwsRestApiResource} awsResource
 * @param {ApiMethod[]} allMethods
 * @returns {Promise}
 *
 */

module.exports.deleteApiResourceMethods = function(apigw, apiTree, awsResource, allMethods) {

  var deleteList;

  try {

    deleteList = awsResource._links['resource:methods'];

    if (deleteList) {

      if (!Array.isArray(deleteList)) deleteList = [deleteList];

      // Filter for awsResource methods that don't exist in apiTree (allMethods)

      deleteList = deleteList.filter(function(method) {
        return allMethods.filter(function(m) {
          return m.method == method.name;
        }).length == 0;
      });

    }

    else deleteList = [];

  } catch (e) {

    debug('awsResource missing _links');
    deleteList = [];

  }

  return Promise.map(deleteList, function(method) {

    return module.exports.deleteApiResourceMethod(apigw, awsResource, method)

  });

}


/**
 * Instance of an AWS Api Gateway RestApi method as returned from AWS
 * 
 * @typedef AwsRestApiMethod
 * @type {object}
 * @property {string} href
 * @property {string} name
 * @property {string} title
 *
 */

/**
 * ### .deleteApiResourceMethod(apigw, apiTree, awsResource, method)
 *
 * Delete specific method from awsResource
 *
 * @api private
 * @param {ApiGatewayApi} apigw
 * @param {AwsRestApiResource} awsResource
 * @param {AwsRestApiMethod} method
 * @returns {Promise}
 *
 */

module.exports.deleteApiResourceMethod = function(apigw, awsResource, method) {

  deleteApiResourceMethod('%j', method);

  return apigw.deleteRestApiMethod(method)

  .then(function() {

    // Remove method from resource listing

    if (Array.isArray(awsResource._links['resource:methods'])) {

      var remove = awsResource._links['resource:methods'].filter(function(m) {
        return method.href != m.href;
      });

      if (remove.length == 1) awsResource._links['resource:methods'] = remove[0];
      else awsResource._links['resource:methods'] = remove;
    }

    else if (awsResource._links['resource:methods'].href == method.href) {
      delete awsResource._links['resource:methods']
    }

  });

}


/**
 * ### .updateApiResourceMethod(apigw, apiTree, awsResource, apiMethod)
 *
 * Updates methods on the Api Gateway with methods from the apiTree
 *
 * @api private
 * @param {ApiGatewayApi} apigw
 * @param {ApiTree} apiTree
 * @param {AwsRestApiResource} awsResource
 * @param {ApiMethod} apiMethod
 * @returns {Promise}
 *
 */

module.exports.updateApiResourceMethod = function(apigw, apiTree, awsResource, apiMethod) {

  // console.log('apiMethod', apiMethod, awsResource);
  updateApiResourceMethod('method: \'%s\', path: \'%s\'', apiMethod.method, awsResource.path);



}
