// aws-deploy/index.js
module.exports = AwsDeploy;

require('./_extends');

var Deployer = require('./lib/deployer');

/**
 * ### .AwsConfig()
 *
 * Configuration object for AwsDeploy
 *
 * @api private
 *
 */

function AwsOpts() {
  this.aws_account = '';
  this.aws_region = '';
  this.aws_access_key_id = '';
  this.aws_secret_access_key = '';
}

/**
 * ### .AwsApiTree()
 *
 * Api definition tree
 *
 * @api private
 *
 */

function AwsApiTree() {/*

  ### example
  
  ```javascript
  {
    name: 'Api Name',
    description: 'Api Description',
    root: {
      _methods: {
        GET: {
          // define integration point for GET /
        }
      }
      'users': {
        _methods: {
          GET: {
            // define integration point for GET /users
          }
        }
        ':id': {
          _methods: {
            GET: {
              // define integration point for GET /users/:id
            }
          }
        }
      }
    }
  }
  ```

*/}

/**
 * ### .AwsDeploy(opts)
 *
 * @api public
 * @param {AwsOpts} opts
 * @param {AwsApiTree} api
 * @returns {Promise}
 *
 */

function AwsDeploy(config, api) {

  return Deployer(config, api);

}
