// aws-deploy/lib/aws/api.js



module.exports = Api;


var Promise = require('bluebird') // available from parent's modules
  , aws4 = require('aws4') // available from parent's modules
  , format = require('util').format
  , https = require('https')
  ;


function Api(config) {
  this.region = config.aws_region;
  this.accessKeyId = config.aws_access_key_id;
  this.secretAccessKey = config.aws_secret_access_key;
  this.host = format('apigateway.%s.amazonaws.com', this.region);
}


Api.prototype.createRequest = function(opts) {
  var headers, date = new Date();

  opts.host = this.host;
  headers = opts.headers = opts.headers || {};
  headers['X-Amz-Date'] = date.toISO8601String();
  headers['Content-Type'] = 'application/x-amz-json-1.0';

  aws4.sign(opts, this);

  return opts;
}


Api.prototype.formatSuccess = function(response) {
  // the array or instance returned
  result = response.data._embedded.item;
  
  // the original http response as hidden property of the result
  Object.defineProperty(result, '_response', {
    value: response
  });

  return result;
}


Api.prototype.formatFailure = function(response) {

  result = {}; // TODO

  Object.defineProperty(result, '_response', {
    value: response
  });
  return result;
}


Api.prototype.performRequest = function(opts) {
  var _this = this;
  return new Promise(function(resolve, reject) {
    https.request(opts, function(res) {

      var response = {
        statusCode: res.statusCode,
        headers: res.headers,
        body: '',
        data: null
      };

      res.on('data', function(data) {
        response.body += data.toString();
      });

      res.on('end', function() {
        if (response.headers['content-type'].match(/json/)) {
          response.data = JSON.parse(response.body);
        }

        if (response.statusCode >= 400) {
          return reject(_this.formatFailure(response));
        }

        if (response.statusCode >= 300) {
          return reject(_this.formatFailure(response));
        }

        if (response.statusCode >= 200) {
          return resolve(_this.formatSuccess(response));
        }

      });

    }).end();
  });
}
