// aws-deploy/lib/aws/api_gateway_api.js
// AWS SDK does not include access to API Gateway

module.exports = ApiGatewayApi;


var Api = require('./api')
  , inherits = require('util').inherits
  , getRestApis = require('debug')('ApiGatewayApi:getRestApis')
  , createRestApi = require('debug')('ApiGatewayApi:createRestApi')
  ;


function ApiGatewayApi(config) {
  Api.call(this, config);
}


inherits(ApiGatewayApi, Api);


/**
 * ### .getRestApis()
 *
 * List all APIs for the given region/account
 *
 * @api public
 * @returns {Promise}
 *
 */

ApiGatewayApi.prototype.getRestApis = function() {

  getRestApis();

  var request = this.createRequest({
    method: 'GET',
    path: '/restapis',
  });

  return this.performRequest(request);

}


/**
 * ### .createRestApi(props)
 *
 * Create a new API
 *
 * @api public
 * @param {{name: String, description: String}} props
 * @returns {Promise}
 *
 */

ApiGatewayApi.prototype.createRestApi = function(props) {

  createRestApi('%j', props);

  var request = this.createRequest({
    method: 'POST',
    path: '/restapis',
    body: JSON.stringify(props),
  });

  return this.performRequest(request);

}
