// aws-deploy/lib/aws/lambda_api.js
module.exports = LambdaApi;

var Api = require('./api')
  , inherits = require('util').inherits
  , getFunction = require('debug')('LambdaApi:getFunction')
  , createFunction = require('debug')('LambdaApi:createFunction')
  , updateFunctionConfiguration = require('debug')('LambdaApi:updateFunctionConfiguration')
  , updateFunctionCode = require('debug')('LambdaApi:updateFunctionCode')
  ;


function LambdaApi(config) {
  Api.call(this, config, 'lambda');
}


inherits(LambdaApi, Api);


/**
 * ### .getFunction(lambda)
 *
 * Gets the AWS function definition for this lambda
 *
 * @api public
 * @param {Lambda} lambda
 * @returns {Promise}
 *
 */

LambdaApi.prototype.getFunction = function(lambda) {

  getFunction('for lambda: \'%s\'', lambda.name);

  var request = this.createRequest({
    method: 'GET',
    path: '/2015-03-31/functions/' + lambda.name,
  });

  return this.performRequest(request);

}


/**
 * ### .createFunction(lambda)
 *
 * Creates the AWS function definition for this lambda
 *
 * @api public
 * @param {Lambda} lambda
 * @returns {Promise}
 *
 */

LambdaApi.prototype.createFunction = function(lambda) {

  createFunction('for lambda: \'%s\'', lambda.name);

  var request = this.createRequest({
    method: 'POST',
    path: '/2015-03-31/functions',
    body: JSON.stringify(lambda.toCreateFunction()),
  });

  return this.performRequest(request);

}


/**
 * ### .updateFunctionConfiguration(lambda)
 *
 * Creates the AWS function definition for this lambda
 *
 * @api public
 * @param {Lambda} lambda
 * @returns {Promise}
 *
 */

LambdaApi.prototype.updateFunctionConfiguration = function(lambda) {

  updateFunctionConfiguration('for lambda: \'%s\'', lambda.name);

  var request = this.createRequest({
    method: 'PUT',
    path: '/2015-03-31/functions/' + lambda.name + '/configuration',
    body: JSON.stringify(lambda.toUpdateFunctionConfiguration()),
  });

  return this.performRequest(request);

}




/**
 * ### .updateFunctionCode(lambda)
 *
 * Creates the AWS function definition for this lambda
 *
 * @api public
 * @param {Lambda} lambda
 * @returns {Promise}
 *
 */

LambdaApi.prototype.updateFunctionCode = function(lambda) {

  updateFunctionCode('for lambda: \'%s\'', lambda.name);

  var request = this.createRequest({
    method: 'PUT',
    path: '/2015-03-31/functions/' + lambda.name + '/code',
    body: JSON.stringify(lambda.toUpdateFunctionCode()),
  });

  return this.performRequest(request);

}
